django/utils/numberformat.py

```diff
@@ -18,46 +18,99 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
-            coefficient, exponent = number.split('e')
-            # Format the coefficient.
-            coefficient = format(
-                coefficient, decimal_sep, decimal_pos, grouping,
-                thousand_sep, force_grouping, use_l10n,
-            )
-            return '{}e{}'.format(coefficient, exponent)
-        else:
-            str_number = '{:f}'.format(number)
+    # Don't unnecessarily convert number to string for speed
+    if isinstance(number, str):
+        return _format_string(number, decimal_sep, decimal_pos, thousand_sep, grouping)
+
+    str_number = str(number)
+    # Treat potentially very large/small floats as Decimals.
+    if isinstance(number, float) and 'e' in str_number.lower():
+        return _format_dec(Decimal(str_number), decimal_sep, decimal_pos, thousand_sep, grouping)
+
+    # Format all other cases as a string
+    return _format_string(str_number, decimal_sep, decimal_pos, thousand_sep, grouping)
+
+
+def _format_int(number, decimal_sep, decimal_pos, thousand_sep, grouping):
+    """
+    Formats an integer, returning it as a string.
+
+    This function should only be used when grouping is 0 or 3. See
+    numberformat.format docstring for details about arguments.
+    """
+    if grouping and (number > 999 or number < -999):
+        # f'{number:,}' returns a string with a ',' as a thousand separator
+        number = f'{number:,}'.replace(',', thousand_sep)
     else:
-        str_number = str(number)
+        # It is unclear as to why this case should be marked safe, but
+        # we return SafeString here to avoid a change in functionality
+        # from the previous implementation of format.
+        number = SafeString(number)
+    if decimal_pos:
+        return number + decimal_sep + '0' * decimal_pos
+    return number
+
+
+def _format_dec(number, decimal_sep, decimal_pos, thousand_sep, grouping):
+    """
+    Gets a number as a Decimal instance, and formats it as a string.
+
+    See numberformat.format docstring for details about arguments.
+    """
+    if decimal_pos is not None:
+        # If the provided number is too small to affect any of the visible
+        # decimal places, consider it equal to '0'.
+        cutoff = Decimal('0.' + '1'.rjust(decimal_pos, '0'))
+        if abs(number) < cutoff:
+            if decimal_pos:
+                return '0' + decimal_sep + '0' * decimal_pos
+            return '0'
+    # Format values with more than 200 digits (an arbitrary cutoff) using
+    # scientific notation to avoid high memory usage in {:f}'.format().
+    _, digits, exponent = number.as_tuple()
+    if abs(exponent) + len(digits) > 200:
+        number = f'{number:e}'
+        coefficient, exponent = number.split('e')
+        # Format the coefficient.
+        coefficient = _format_string(
+            coefficient, decimal_sep, decimal_pos,
+            thousand_sep, grouping,
+        )
+        return f'{coefficient}e{exponent}'
+
+    number = f'{number:f}'
```

Comment:
Could be a personal or project preference for all I know, but does this re-assignment need to occur, instead of just having `return _format_string(f'{number:f}', ...)`? Doubt it matters much, of course.